File: hosts/go/engine/session.go
package engine

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sync"

	"github.com/fogha/teml/hosts/go/ndjson"
	"github.com/fogha/teml/hosts/go/protocol"
)

// SpawnOptions configures a teml run child process.
type SpawnOptions struct {
	ResolveOptions
	ViewPath string
	Width    int
	Height   int
	// Frames and Mode are optional CLI flags. Leave empty when negotiating via
	// stdin configure — do not combine startup --frames/--mode with configure.
	Frames  protocol.FrameFormat
	Mode    protocol.FrameMode
	NoColor bool
}

// Session drives teml run over piped NDJSON stdio with unbuffered pipe I/O.
type Session struct {
	child    *exec.Cmd
	stdinRaw io.WriteCloser
	stdin    *bufio.Writer
	stdout   io.Reader
	splitter ndjson.Splitter
	pending  []ndjson.Input
	mu       sync.Mutex
	closed   bool
	Engine   ResolvedEngine
}

// Spawn starts teml run with the resolved engine and returns after pipes are ready.
func Spawn(opts SpawnOptions) (*Session, error) {
	engine, err := Resolve(opts.ResolveOptions)
	if err != nil {
		return nil, err
	}
	if opts.ViewPath == "" {
		return nil, errors.New("view path is required")
	}
	width := opts.Width
	if width < 1 {
		width = 80
	}
	height := opts.Height
	if height < 1 {
		height = 24
	}
	args := append(append([]string{}, engine.Args...),
		"run", opts.ViewPath,
		"--width", fmt.Sprintf("%d", width),
		"--height", fmt.Sprintf("%d", height),
	)
	if opts.Frames != "" {
		args = append(args, "--frames", string(opts.Frames))
	}
	if opts.Mode != "" {
		args = append(args, "--mode", string(opts.Mode))
	}
	if opts.NoColor {
		args = append(args, "--no-color")
	}

	cmd := exec.Command(engine.Program, args...)
	cmd.Stderr = os.Stderr
	stdinPipe, err := cmd.StdinPipe()
	if err != nil {
		return nil, fmt.Errorf("teml stdin pipe: %w", err)
	}
	stdoutPipe, err := cmd.StdoutPipe()
	if err != nil {
		_ = stdinPipe.Close()
		return nil, fmt.Errorf("teml stdout pipe: %w", err)
	}
	if err := cmd.Start(); err != nil {
		_ = stdinPipe.Close()
		_ = stdoutPipe.Close()
		return nil, fmt.Errorf("spawn teml engine (%s): %w", engine.Source, err)
	}

	return &Session{
		child:    cmd,
		stdinRaw: stdinPipe,
		stdin:    bufio.NewWriter(stdinPipe),
		stdout:   stdoutPipe,
		Engine:   engine,
	}, nil
}

// Send writes one command as a single NDJSON line and flushes immediately.
func (s *Session) Send(cmd protocol.Command) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closed {
		return errors.New("session closed")
	}
	line, err := cmd.MarshalNDJSON()
	if err != nil {
		return err
	}
	if len(line) > protocol.MaxNDJSONLineBytes {
		return fmt.Errorf("command exceeds the %d-byte NDJSON limit", protocol.MaxNDJSONLineBytes)
	}
	if _, err := s.stdin.Write(line); err != nil {
		return fmt.Errorf("write command: %w", err)
	}
	if err := s.stdin.WriteByte('\n'); err != nil {
		return fmt.Errorf("write command newline: %w", err)
	}
	if err := s.stdin.Flush(); err != nil {
		return fmt.Errorf("flush command: %w", err)
	}
	return nil
}

// Next reads the next non-blank session event.
func (s *Session) Next() (protocol.SessionEvent, error) {
	for {
		line, err := s.readLine()
		if err != nil {
			return protocol.SessionEvent{}, err
		}
		return protocol.DecodeEvent(line)
	}
}

// NextFrame skips semantic events until a frame arrives.
func (s *Session) NextFrame() (protocol.SessionEvent, error) {
	for {
		event, err := s.Next()
		if err != nil {
			return protocol.SessionEvent{}, err
		}
		switch event.Type {
		case "frame":
			return event, nil
		case "error":
			return protocol.SessionEvent{}, fmt.Errorf("protocol error: %s", event.Message)
		case "exit":
			return protocol.SessionEvent{}, errors.New("session exited before producing a frame")
		}
	}
}

// InitialFrame reads the mandatory first frame event.
func (s *Session) InitialFrame() (protocol.SessionEvent, error) {
	event, err := s.Next()
	if err != nil {
		return protocol.SessionEvent{}, err
	}
	if event.Type != "frame" {
		return protocol.SessionEvent{}, fmt.Errorf("protocol violation: expected initial frame, got %q", event.Type)
	}
	return event, nil
}

// Close terminates the child process and closes pipes.
func (s *Session) Close() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closed {
		return nil
	}
	s.closed = true
	_ = s.stdin.Flush()
	_ = s.stdinRaw.Close()
	if s.child.Process != nil {
		_ = s.child.Process.Kill()
	}
	err := s.child.Wait()
	if err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			return nil
		}
		return err
	}
	return nil
}

// Wait waits for the child to exit after a clean exit command.
func (s *Session) Wait() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.child.Process == nil {
		return errors.New("session not started")
	}
	if !s.closed {
		_ = s.stdin.Flush()
		_ = s.stdinRaw.Close()
		s.closed = true
	}
	return s.child.Wait()
}

func (s *Session) readLine() (string, error) {
	if line, err, ok := s.popPending(); ok {
		return line, err
	}

	chunk := make([]byte, 4096)
	for {
		n, err := s.stdout.Read(chunk)
		if n > 0 {
			s.pending = append(s.pending, s.splitter.Push(string(chunk[:n]))...)
			if line, pendingErr, ok := s.popPending(); ok {
				return line, pendingErr
			}
		}
		if err != nil {
			if errors.Is(err, io.EOF) {
				s.pending = append(s.pending, s.splitter.Flush()...)
				if line, pendingErr, ok := s.popPending(); ok {
					return line, pendingErr
				}
				return "", errors.New("teml closed stdout")
			}
			return "", err
		}
	}
}

func (s *Session) popPending() (string, error, bool) {
	if len(s.pending) == 0 {
		return "", nil, false
	}
	input := s.pending[0]
	s.pending = s.pending[1:]
	if input.Error != "" {
		return "", errors.New(input.Error), true
	}
	return input.Line, nil, true
}

File: crates/teml-host/src/session.rs
//! NDJSON session over a spawned `teml run` child process.

use crate::engine::{
    default_package_scripts, resolve_engine, EngineError, EngineResolveOptions, ResolvedEngine,
};
use crate::protocol::{Command, Event, Frame, FrameFormat, FrameMode};
use std::fmt;
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};
use std::process::{Child, ChildStdin, ChildStdout, Command as ProcessCommand, Stdio};

/// Startup options for [`Session::spawn`].
#[derive(Debug, Clone)]
pub struct SessionOptions {
    /// Document path passed to `teml run`.
    pub document: PathBuf,
    /// Initial terminal width in columns.
    pub width: u16,
    /// Initial terminal height in rows (`None` omits `--height`).
    pub height: Option<u16>,
    /// Pre-negotiated frame payload format (`None` omits `--frames`).
    pub frames: Option<FrameFormat>,
    /// Pre-negotiated frame delivery mode (`None` omits `--mode`).
    pub mode: Option<FrameMode>,
    /// Disable ANSI color in engine output.
    pub no_color: bool,
    /// Engine resolution overrides.
    pub engine: EngineResolveOptions,
}

impl SessionOptions {
    /// Convenience constructor with common defaults for interactive hosts.
    pub fn new(document: impl Into<PathBuf>, width: u16, height: u16) -> Self {
        Self {
            document: document.into(),
            width,
            height: Some(height),
            frames: Some(FrameFormat::Ansi),
            mode: Some(FrameMode::Patches),
            no_color: false,
            engine: EngineResolveOptions::default(),
        }
    }

    /// Populate default monorepo package script search paths from a manifest dir.
    pub fn with_default_package_scripts(mut self, manifest_dir: &Path) -> Self {
        self.engine.package_scripts = default_package_scripts(manifest_dir);
        self
    }
}

/// Active `teml run` subprocess speaking NDJSON over piped stdio.
pub struct Session {
    child: Child,
    stdin: ChildStdin,
    stdout: BufReader<ChildStdout>,
    engine: ResolvedEngine,
    closed: bool,
}

impl Session {
    /// Spawn `teml run` with explicit engine resolution and startup flags.
    pub fn spawn(options: SessionOptions) -> Result<Self, SessionError> {
        let engine = resolve_engine(&options.engine)?;

        let mut command = ProcessCommand::new(&engine.program);
        command
            .args(&engine.args)
            .arg("run")
            .arg(&options.document)
            .arg("--width")
            .arg(options.width.max(1).to_string());
        if let Some(height) = options.height {
            command.arg("--height").arg(height.max(1).to_string());
        }
        if let Some(frames) = options.frames {
            command.arg("--frames").arg(frame_format_arg(frames));
        }
        if let Some(mode) = options.mode {
            command.arg("--mode").arg(frame_mode_arg(mode));
        }
        if options.no_color {
            command.arg("--no-color");
        }
        command
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::inherit());

        let mut child = command.spawn().map_err(SessionError::Io)?;
        let stdin = child
            .stdin
            .take()
            .ok_or_else(|| SessionError::Protocol("teml stdin pipe was not created".into()))?;
        let stdout = child
            .stdout
            .take()
            .ok_or_else(|| SessionError::Protocol("teml stdout pipe was not created".into()))?;

        Ok(Self {
            child,
            stdin,
            stdout: BufReader::new(stdout),
            engine,
            closed: false,
        })
    }

    /// Resolved engine metadata and version diagnostics.
    pub fn engine(&self) -> &ResolvedEngine {
        &self.engine
    }

    /// Read the first frame emitted before any command (startup contract).
    pub fn initial_frame(&mut self) -> Result<Frame, SessionError> {
        match self.next_event()? {
            Event::Frame(frame) => Ok(frame),
            other => Err(SessionError::Protocol(format!(
                "protocol violation: expected initial frame, got {other:?}"
            ))),
        }
    }

    /// Write one command as a single NDJSON line.
    pub fn send(&mut self, command: &Command) -> Result<(), SessionError> {
        let line = serde_json::to_string(command)
            .map_err(|error| SessionError::Protocol(error.to_string()))?;
        writeln!(self.stdin, "{line}").map_err(SessionError::Io)?;
        self.stdin.flush().map_err(SessionError::Io)
    }

    /// Read the next event from stdout, skipping blank lines.
    pub fn next_event(&mut self) -> Result<Event, SessionError> {
        let mut line = String::new();
        loop {
            line.clear();
            let read = self.stdout.read_line(&mut line).map_err(SessionError::Io)?;
            if read == 0 {
                return Err(SessionError::Protocol("teml closed stdout".into()));
            }
            let trimmed = line.trim();
            if !trimmed.is_empty() {
                return Event::from_line(trimmed).map_err(|error| {
                    SessionError::Protocol(format!("invalid event JSON: {error}"))
                });
            }
        }
    }

    /// Drain events until the next `frame` (surfacing protocol errors immediately).
    pub fn next_frame(&mut self) -> Result<Frame, SessionError> {
        loop {
            match self.next_event()? {
                Event::Frame(frame) => return Ok(frame),
                Event::Error { message } => {
                    return Err(SessionError::Protocol(format!("protocol error: {message}")));
                }
                Event::Exit => {
                    return Err(SessionError::Protocol(
                        "session exited before producing a frame".into(),
                    ));
                }
                Event::Change { .. }
                | Event::Toggle { .. }
                | Event::Click { .. }
                | Event::Unknown => {}
            }
        }
    }

    /// Send `exit` and wait for the child, returning its exit status.
    pub fn close(mut self) -> Result<std::process::ExitStatus, SessionError> {
        if !self.closed {
            let _ = self.send(&Command::Exit);
            self.closed = true;
        }
        let status = self.child.wait().map_err(SessionError::Io)?;
        Ok(status)
    }
}

impl Drop for Session {
    fn drop(&mut self) {
        if !self.closed {
            let _ = self.send(&Command::Exit);
            let _ = self.child.kill();
            let _ = self.child.wait();
        }
    }
}

/// Session lifecycle and transport failures.
#[derive(Debug)]
pub enum SessionError {
    /// Operating-system I/O while talking to the child process.
    Io(std::io::Error),
    /// Engine resolution failed before spawn.
    Engine(EngineError),
    /// NDJSON transport or protocol invariant violation.
    Protocol(String),
}

impl fmt::Display for SessionError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Io(error) => write!(f, "{error}"),
            Self::Engine(error) => write!(f, "{error}"),
            Self::Protocol(message) => write!(f, "{message}"),
        }
    }
}

impl std::error::Error for SessionError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Self::Io(error) => Some(error),
            Self::Engine(error) => Some(error),
            Self::Protocol(_) => None,
        }
    }
}

impl From<std::io::Error> for SessionError {
    fn from(value: std::io::Error) -> Self {
        Self::Io(value)
    }
}

impl From<EngineError> for SessionError {
    fn from(value: EngineError) -> Self {
        Self::Engine(value)
    }
}

fn frame_format_arg(format: FrameFormat) -> &'static str {
    match format {
        FrameFormat::Ansi => "ansi",
        FrameFormat::Plain => "plain",
        FrameFormat::Both => "both",
    }
}

fn frame_mode_arg(mode: FrameMode) -> &'static str {
    match mode {
        FrameMode::Full => "full",
        FrameMode::Patches => "patches",
    }
}
